File: 1000 and 1 tasks/leetcode-75-2/minimum-window-substring.cs
public class Solution {
    private bool IsValid(string s, string t) {
        var tableLower = new int[26];
        var tableUpper = new int[26];

        foreach (var c in s) {
            if (Char.IsUpper(c)) {
                tableUpper[c - 'A']++;
            }
            else {
                tableLower[c - 'a']++;
            }
        }

        foreach (var c in t) {
            if (Char.IsUpper(c)) {
                tableUpper[c - 'A']--;
                if (tableUpper[c - 'A'] < 0) {
                    return false;
                }
            }
            else {
                tableLower[c - 'a']--;
                if (tableLower[c - 'a'] < 0) {
                    return false;
                }
            }
        }

        return true;
    }

    public string MinWindow(string s, string t) {
        var length = t.Length;
        while (length <= s.Length) {
            var possibleStartCount = s.Length - length + 1;
            for (int i = 0; i < possibleStartCount; i++) {
                var sub = s.Substring(i, length);
                if (IsValid(sub, t)) {
                    return sub;
                }
            }
            length++;
        }

        return "";
    }
}

File: 1000 and 1 tasks/new-top-150/minimum-window-substring.java
import java.util.*;

public class Solution {

    public String minWindow(String s, String t) {
        var tMap = new HashMap<Character, Integer>();
        for (var c : t.toCharArray()) {
            tMap.put(c, tMap.getOrDefault(c, 0) + 1);
        }

        String result = "";
        var min = Integer.MAX_VALUE;

        for (var left = 0; left < s.length(); left++) {
            for (var right = left; right < s.length(); right++) {
                var sub = s.substring(left, right + 1);

                var isValid = true;
                for (var key : tMap.keySet()) {
                    var count = tMap.get(key);
                    if (sub.chars().filter(c -> c == key).count() < count) {
                        isValid = false;
                        break;
                    }
                }
                if (isValid && sub.length() < min) {
                    min = sub.length();
                    result = sub;
                }
            }
        }
        return result;
    }
}